core/src/main/java/org/mule/runtime/core/api/retry/policy/RetryPolicyTemplate.java

```diff
@@ -75,11 +96,35 @@
    * @param <T> the generic type of the publisher's content
    * @return a {@link Publisher} configured with the retry policy.
    * @since 4.0
+   *
+   * @deprecated Use {@link #applyPolicy(Publisher, Predicate, Consumer, Function, Optional)} instead
    */
+  @Deprecated
   default <T> Publisher<T> applyPolicy(Publisher<T> publisher,
                                        Predicate<Throwable> shouldRetry,
                                        Consumer<Throwable> onExhausted, Function<Throwable, Throwable> errorFunction) {
-    return createRetryInstance().applyPolicy(publisher, shouldRetry, onExhausted, errorFunction);
+    return applyPolicy(publisher, shouldRetry, onExhausted, errorFunction, empty());
+
+  }
+
+  /**
+   * Applies the retry policy in a non blocking manner by transforming the given {@code publisher} into one configured to apply
+   * the retry logic.
+   *
+   * @param publisher a publisher with the items which might fail
+   * @param shouldRetry a predicate which evaluates each item to know if it should be retried or not
+   * @param onExhausted an action to perform when the retry action has been exhausted
+   * @param errorFunction function used to map cause exception to exception emitted by retry policy.
+   * @param retryScheduler the scheduler to use when retrying. If empty, an internal reactor Scheduler will be used.
+   * @param <T> the generic type of the publisher's content
+   * @return a {@link Publisher} configured with the retry policy.
+   * @since 4.2
+   */
+  default <T> Publisher<T> applyPolicy(Publisher<T> publisher,
+                                       Predicate<Throwable> shouldRetry,
+                                       Consumer<Throwable> onExhausted, Function<Throwable, Throwable> errorFunction,
+                                       Optional<Scheduler> retryScheduler) {
```

Comment:
I don't think that any of these schedulers should be optional. I'm putting this comment here but this applies PR wide